src/lxc/conf.c

```diff
@@ -3930,29 +3940,60 @@ static void turn_into_dependent_mounts(const struct lxc_rootfs *rootfs)
 	 */
 	move_fd(memfd);
 
+	char p_mount[PATH_MAX], p_path[PATH_MAX];
+	bool p_mount_shared, p_path_shared;
+
+	p_mount[0] = '\0';
+	p_path[0] = '\0';
+
 	while (getline(&line, &len, f) != -1) {
 		char *opts, *target;
+		bool shared;
+
+		TRACE("mountinfo: %s", line);
 
 		target = get_field(line, 4);
 		if (!target)
 			continue;
 
 		opts = get_field(target, 2);
-		if (!opts)
-			continue;
-
 		null_endofword(opts);
-		if (!strstr(opts, "shared"))
-			continue;
+
+		if (opts && strstr(opts, "shared"))
+			shared = true;
 
 		null_endofword(target);
-		ret = mount(NULL, target, NULL, MS_SLAVE, NULL);
+
+		// Find parent mount of rootfs->path (longest prefix match)
+		if ((strnequal(rootfs->path, target, strlen(target))) &&
+		    (strlen(target) > strlen(p_path))) {
+			(void)strlcpy(p_path, target, sizeof(p_path));
+			p_path_shared = shared;
+		}
+
+		// Find parent mount of rootfs->mount (longest prefix match)
+		if ((strnequal(rootfs->mount, target, strlen(target))) &&
+		    (strlen(target) > strlen(p_mount))) {
+			(void)strlcpy(p_mount, target, sizeof(p_mount));
+			p_mount_shared = shared;
+		}
+	}
+
+	if (p_path_shared) {
+		DEBUG("Change mount propagation for \"%s\" to private (lxc.rootfs.path parent)", p_path);
+		ret = mount(NULL, p_path, NULL, MS_SLAVE, NULL);
 		if (ret < 0) {
-			SYSERROR("Failed to recursively turn old root mount tree into dependent mount. Continuing...");
-			continue;
+			SYSERROR("Failed to change mount propagation for \"%s\" to private (lxc.rootfs.path parent)", p_path);
+		}
+	}
+
+	if (p_mount_shared) {
+		DEBUG("Change mount propagation for \"%s\" to private (lxc.rootfs.mount parent)", p_mount);
+		ret = mount(NULL, p_mount, NULL, MS_SLAVE, NULL);
+		if (ret < 0) {
+			SYSERROR("Failed to change mount propagation for \"%s\" to private (lxc.rootfs.mount parent)", p_mount);
 		}
 	}
-	TRACE("Turned all mount table entries into dependent mount");
 }
```

Comment:
just to ensure that I understand your idea right:
you are changing the `turn_into_dependent_mounts` function such that it stops setting `MS_SLAVE` on all container mounts, but only for `rootfs->path` and `rootfs->mount` parent mounts, right?

Reply:
@mihalicyn Yes that was my intention. I tried to describe it in the commit message of  https://github.com/lxc/lxc/pull/3880/commits/3f85b34bc73de539174c6bcda43150cbf64a3539 as best as I can.
I'll have to look for my mount propagation test scripts - they should be a start point for implementing lxc tests.

I share all of your points in the review. But since I've switched to runc as container runtime and paused development of https://github.com/lxc/lxcri I really wonder if anyone benefits from this PR ?

Reply:
To be honest I'm not very familiar with the lxcri project ;-)
But I guess that if this issue is blocking lxcri from functioning properly we have to do something with this.

>Yes that was my intention. I tried to describe it in the commit message of https://github.com/lxc/lxc/commit/3f85b34bc73de539174c6bcda43150cbf64a3539 as best as I can.

Your description was great, but I just wanted to ensure and sync our understanding of the problem and approach, because as you've mentioned all of this is very sensitive stuff and can cause a security issues. That's why it's better to ask a few extra questions than having any doubts :)

>But since I've switched to runc as container runtime and paused development

Ah. You was using the lxcri with Kubernetes, right?